### PR TITLE
feature: Use `evmone-t8n` to fill tests

### PR DESCRIPTION
**Pull request by marioevz**
Use https://github.com/ethereum/evmone to fill tests.

Builds on top of work started by @spencer-tb.

Also implements transaction signing for all transaction types.

### CODE REVIEW

**Comment by marioevz** [comment]
@chfast @winsvega Evmone it's almost fully working now!

To run all tests using evmone-t8n use this branch and run:
```bash
fill --evm-bin=evmone-t8n
```
`evmone-t8n` should be available in path for it to work.

All but two pre-Cancun tests are filling correctly:
- tests/shanghai/eip4895_withdrawals/test_withdrawals.py::test_zero_amount[fork=Shanghai-four_withdrawals_one_with_value_one_with_max]
- tests/shanghai/eip4895_withdrawals/test_withdrawals.py::test_zero_amount[fork=Shanghai-four_withdrawals_one_with_value_one_with_max_reversed_order]

It seems like an account with an zero-value withdrawal remains in the state.

To run:
```bash
fill tests/shanghai/eip4895_withdrawals/test_withdrawals.py::test_zero_amount[fork=Shanghai-four_withdrawals_one_with_value_one_with_max] --evm-bin=evmone-t8n
```
and
```bash
fill tests/shanghai/eip4895_withdrawals/test_withdrawals.py::test_zero_amount[fork=Shanghai-four_withdrawals_one_with_value_one_with_max_reversed_order] --evm-bin=evmone-t8n
```

**Review by danceratopz — COMMENTED** [review]
This is great! Love how easy it is to add a new transition tool.

I would suggest we split `src/evm_transition_tool/__init__.py` into multiple files. Typically, we've only defined the module/package namespaces in our init files; I think this is the only init that defines objects within it.

```
➜ find src -name __init__.py -exec wc -c '{}' \;
1270 src/ethereum_test_forks/__init__.py
50 src/ethereum_test_forks/tests/__init__.py
53 src/ethereum_test_forks/forks/__init__.py
467 src/ethereum_test_tools/spec/__init__.py
287 src/ethereum_test_tools/reference_spec/__init__.py
313 src/ethereum_test_tools/code/__init__.py
84 src/ethereum_test_tools/filling/__init__.py
1709 src/ethereum_test_tools/__init__.py
1520 src/ethereum_test_tools/common/__init__.py
50 src/ethereum_test_tools/tests/__init__.py
148 src/ethereum_test_tools/vm/__init__.py
186 src/pytest_plugins/spec_version_checker/__init__.py
86 src/pytest_plugins/test_filler/__init__.py
67 src/pytest_plugins/__init__.py
167 src/pytest_plugins/forks/__init__.py
13592 src/evm_transition_tool/__init__.py  <--------------------------------------------------------- !
```

How about moving:
- `TransitionTool` into `transition_tool.py`
- `EvmTransitionTool` into `geth.py` (and perhaps rename the class to `GethTransitionTool`
- EvmOneTransitionTool  into `evmone.py`
?

Otherwise, couple of minor comments below.

**Comment by marioevz** [comment]
@danceratopz Thanks for the review! :)

Thanks to this handy stackoverflow response https://stackoverflow.com/a/66883553, I've added a dynamic method to load all transition tools automatically. Check it out: https://github.com/marioevz/execution-spec-tests/blob/3a28c9c50d3c905d39add0c7c2e12974d0043d22/src/evm_transition_tool/transition_tool.py#L37

**Comment by danceratopz** [comment]
> Thanks to this handy stackoverflow response https://stackoverflow.com/a/66883553, I've added a dynamic method to load all transition tools automatically. Check it out: https://github.com/marioevz/execution-spec-tests/blob/3a28c9c50d3c905d39add0c7c2e12974d0043d22/src/evm_transition_tool/transition_tool.py#L37

This is awesome. I went a step further and modified `from_binary()` to return the instantiated class directly. Then I cleaned up the types used for `binary`/`binary_path` and moved some common code to the base class. Hope that's all right?! :pray:

**Comment by marioevz** [comment]
> Just one comment. `matches_binary_path` doesn't seem like a very robust method. It looks like multiple tools call their binary `evm`. Ideally, they'd have a `--name` flag or similar. Or perhaps we can use the `-v` option or ask teams to add a unique "id" there?

I've harden the tool detection in https://github.com/ethereum/execution-spec-tests/pull/142/commits/622f181552642c2cb1fb48f9c0552644b81cb974.

Turns out, all current evm tools handle `-v` flag, so we can simply regex the result and detect the correct tool.

File: src/evm_transition_tool/transition_tool.py [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+                    binary_output = os.popen(f"{binary} {version_flag}").read()
+                except Exception:
+                    # If the tool doesn't support the version flag,
+                    # we'll get an non-zero exit code.
+                    continue
+                for subclass in subclasses:
+                    if subclass.detect_binary(binary_output):
+                        return subclass(binary=binary, **kwargs)
+
+        raise UnknownTransitionTool(f"Unknown transition tool binary: {binary_path}")
```
Comment (danceratopz) [inline src/evm_transition_tool/transition_tool.py:123]: I think this is the appropriate error handling:
```suggestion
                raise UnknownTransitionTool(f"Unknown transition tool binary: {binary_path}")

        raise TransitionToolNotFoundInPath(binary=binary)
```

File: src/evm_transition_tool/transition_tool.py [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+        for version_flag, subclasses in groupby(
+            cls.registered_tools, key=lambda x: x.version_flag
+        ):
+            try:
+                with os.popen(f"{binary} {version_flag}") as f:
+                    binary_output = f.read()
+            except Exception:
+                # If the tool doesn't support the version flag,
+                # we'll get an non-zero exit code.
+                continue
```
[the remark this answers is not in the document]
Comment (danceratopz) [reply src/evm_transition_tool/transition_tool.py:122]: Something to keep in mind for the future. Non-blocking for this PR imo :)

[the remark this answers is not in the document]
Comment (marioevz) [reply src/evm_transition_tool/transition_tool.py:122]: I think, e.g. if some tool changes the output of this flag, we would start getting `UnknownTransitionTool` exceptions on all tests, and we could take a look and make the appropriate fix then.

**Comment by chfast** [comment]
This should fix it (but untested): https://github.com/ethereum/evmone/pull/670.

Is this enough to `fill --evm-bin=evmone-t8n` to get test failure reports? If yes this should be good enough for us as the way of running these tests.

**Comment by danceratopz** [comment]
Hi @chfast,

> This should fix it (but untested): [ethereum/evmone#670](https://github.com/ethereum/evmone/pull/670).

Nice! I confirmed that the errors are indeed fixed in your branch (0.11.0-dev+commit.436f5f18).

> Is this enough to `fill --evm-bin=evmone-t8n` to get test failure reports? If yes this should be good enough for us as the way of running these tests.

Yes, if `evmone-t8n` is in your path. Otherwise you can also provide an absolute or relative path to the `evmone-t8n` binary.

Just to clarify whether it's good enough, I'm sure you're aware, but `fill` only checks the limited post conditions defined in the test, in this case https://github.com/ethereum/execution-spec-tests/blob/ace89a1e4af4394f1d1e3346fa5d5f43cb6c6edf/tests/shanghai/eip4895_withdrawals/test_withdrawals.py#L638; it's not considered the full test (as defined by the corresponding JSON fixture that it generates).

**Comment by danceratopz** [comment]
Hi @chfast,

> This should fix it (but untested): [ethereum/evmone#670](https://github.com/ethereum/evmone/pull/670).

Nice! I confirmed that the errors are indeed fixed in your branch (0.11.0-dev+commit.436f5f18).

> Is this enough to `fill --evm-bin=evmone-t8n` to get test failure reports? If yes this should be good enough for us as the way of running these tests.

Yes, if `evmone-t8n` is in your path. Otherwise you can also provide an absolute or relative path to the `evmone-t8n` binary.

Just to clarify as to whether it's good enough, as I'm sure you're aware, `fill` only checks the limited post conditions defined in the test, in this case https://github.com/ethereum/execution-spec-tests/blob/ace89a1e4af4394f1d1e3346fa5d5f43cb6c6edf/tests/shanghai/eip4895_withdrawals/test_withdrawals.py#L638
It's not considered the full test (as defined by the corresponding JSON fixture that it generates).

**Comment by chfast** [comment]
For the reference, I can run it as described. It works very good (a lot of failures to fix) however debugging is difficult.

**Pull request merged**